### ISSUE TITLE
rewrite TestUnderTemporaryNetworkFailure to use nftables

### ISSUE THREAD

**Issue by danwinship** [kubernetes/kubernetes#119047]
Some e2e tests use [`TestUnderTemporaryNetworkFailure` in `test/e2e/framework/network`](https://github.com/kubernetes/kubernetes/blob/v1.27.0/test/e2e/framework/network/utils.go#L1078) to test how a component behaves when the network goes down. (Additionally, [one test in `test/e2e/apimachinery`](https://github.com/kubernetes/kubernetes/blob/v1.27.0/test/e2e/apimachinery/etcd_failure.go#L63) does the same thing "by hand" but ought to be rewritten to use `TestUnderTemporaryNetworkFailure`, or at least to use the `BlockNetwork` / `UnblockNetwork` functions that it uses.)

This test currently uses iptables rules to simulate the network failure, but in the future, nftables will be more widely-supported/widely-available than iptables, so it should be rewritten to try using `nft` first, and only fall back to `iptables` if `nft` isn't installed.

/sig network
/help-wanted

**Comment by shaneutt** [comment]
We talked about this in the SIG Network sync, it's as straightforward as it sounds and we want to do this one:

/triage accepted

**Comment by mansi1597** [comment]
I would like to work on this.
/assign

**Comment by uablrek** [comment]
Is there and idea/guideline/poc on how to implement the interaction with nftables? The main alternatives as I see it are:

1. Call the `nft` program
2. Go "directly" to nftables

The later involves some very low-level configuration, likely using https://github.com/google/nftables. The former is made easier by the `--json` option to `nft`.

I would suggest that an API is defined that allows either implementation.

Please note that operations towards nftables are atomic. This should be reflected in the API, e.g. by a `commit()` operation. A number of [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) operations build some structure beneath the API and the `commit()` sends the commands as one operation to nftables (via `nft` or direct). The `commit()` may fail, and that must be taken care of, of course.

**Comment by uablrek** [comment]
The "commit" function is called [Flush()](https://pkg.go.dev/github.com/google/nftables#Conn.Flush in the google API.

**Comment by uablrek** [comment]
There is an implementation in `rust` at https://docs.rs/nftables-json/latest/nftables_json/. I haven't found any in `go`.

Manual test:
```
nft --json -f test1.json
nft --json list ruleset | jq
```

<details><summary>test1.json</summary>

```json
{"nftables":[
  {"flush":{"ruleset":null}},
  {"add":{"table":{"family":"inet","name":"default"}}},
  {"add":{"chain":{"family":"inet","table":"default","name":"input","policy":"accept","type":"filter","hook":"input","prio":0}}},
  {"add":{"rule":{"family":"inet","table":"default","chain":"input","expr":[
    {"match":{"left":{"payload":{"protocol":"udp","field":"sport"}},"right":53,"op":"=="}},
    {"drop":null}
  ]}}}
]}
```

</details>

<details><summary>output</summary>

```json
{
  "nftables": [
    {
      "metainfo": {
        "version": "1.0.2",
        "release_name": "Lester Gooch",
        "json_schema_version": 1
      }
    },
    {
      "table": {
        "family": "inet",
        "name": "default",
        "handle": 2
      }
    },
    {
      "chain": {
        "family": "inet",
        "table": "default",
        "name": "input",
        "handle": 1,
        "type": "filter",
        "hook": "input",
        "prio": 0,
        "policy": "accept"
      }
    },
    {
      "rule": {
        "family": "inet",
        "table": "default",
        "chain": "input",
        "handle": 2,
        "expr": [
          {
            "match": {
              "op": "==",
              "left": {
                "payload": {
                  "protocol": "udp",
                  "field": "sport"
                }
              },
              "right": 53
            }
          },
          {
            "drop": null
          }
        ]
      }
    }
  ]
}
```

</details>

**Comment by danwinship** [comment]
> Is there any idea/guideline/poc on how to implement the interaction with nftables?

So for purposes of _this_ issue, I think the test function should just call `nft` directly, without using any sort of library/package/abstraction. (The existing TestUnderTemporaryNetworkFailure implementation just calls `iptables` directly.) It only needs to add a few rules. It's not worth doing anything more complicated.

For the kube-proxy backend, I'm planning to use https://github.com/danwinship/nftables, but that discussion belongs on the KEP PR, not here.

**Comment by mansi1597** [comment]
If we first try blocking using nft table rule and then use iptables as fallback option, while unblocking, how can we figure out which of the 2- iptable or nftable rule needs to be dropped. Instead of doing this, should we first check if nft is installed and enabled, and based on that call the unblock function?

**Comment by mansi1597** [comment]
@danwinship Can you please review the PR linked to close this issue?

**Comment by aojea** [comment]
> but in the future, nftables will be more widely-supported/widely-available than iptables, 

I will be more conservative here, iptables is everywhere and nftables not, can we be more precise about the future to set expectactions?

**Comment by danwinship** [comment]
> iptables is everywhere and nftables not

not true! At least Debian switched from iptables being a required package to nftables being a required package (with iptables _almost certainly_ getting pulled in as a dependency of something else, but still).

Anyway, I have no expected timeline. We can do iptables-with-fallback-to-nftables if you prefer. It doesn't really matter much either way.

**Comment by sadB0nk** [comment]
@mansi1597 are you still working on it?
